Python/ceval.c

```diff
@@ -5022,12 +5022,23 @@ import_from(PyObject *v, PyObject *name)
     return x;
  error:
     pkgpath = PyModule_GetFilenameObject(v);
+    if (pkgname == NULL) {
+        pkgname_or_unknown = PyUnicode_FromString("<unknown module name>");
+    } else {
+        pkgname_or_unknown = pkgname;
+    }
```

Comment:
I think there are some reference counting bugs lurking here.

* It's possible the code jumps to `error:` after `pkgname` is decref'd.  I think that decref should be deferred until just before the good-path return.
* `pkgname_or_unknown` will be a new reference if `pkgname` is NULL, or it will steal the `pkgname` reference if it's not.  In either case, the `error` stanza should decref `pkgname_or_unknown`, which will take care of ensuring the object gets decref for any `goto error` path.
* `pkgpath` doesn't get decref'd.  I think you should `Py_XDECREF` it before the `return NULL` just in case `pkgpath` is itself NULL.